### PR TITLE
SCP-4799 - Fix JsonStream bug

### PR DESCRIPTION
**Pull request by hrajchert**
This PR fixes some bugs in the Control.JsonStream module that didn't dispatch `onJson` in certain situations and that didn't recover well from errors. More details are in the comment.

It also adds a `Co`rutine wrapper to unit test the module

### CODE REVIEW

File: flake.lock [hunk 1]
```diff
@@ -86,11 +86,11 @@
...
     "nixpkgs": {
       "locked": {
-        "lastModified": 1660227034,
-        "narHash": "sha256-bXMlG/YU0IjAod6M625XT1YbUG+/3L9ypk9llYpKeuM=",
+        "lastModified": 1670498129,
+        "narHash": "sha256-QOWbsKObHxNpBzrIVJrgR2TGQdwyYNFgBdoVCCvYuDU=",
         "owner": "NixOS",
         "repo": "nixpkgs",
-        "rev": "964d60ff2e6bc76c0618962da52859603784fa78",
+        "rev": "52665b05cfd702e12b877aa519b62374d003231a",
```
Comment (hrajchert) [inline flake.lock:93]: Updated nixpkgs to have the latest version of node

File: flake.nix [hunk 2]
```diff
@@ -170,6 +170,7 @@
               test
               prettier
               bower
+              nodejs
```
Comment (hrajchert) [inline flake.nix:173]: Added nodejs to the shell to make sure that we have the same version. I was running into an issue of nodejs buffer methods not being available in my local version

File: package.json [hunk 3]
```diff
@@ -7,7 +7,7 @@
     "build": "spago build",
     "build-spec": "spago -x spec-test.dhall build",
     "run-spec": "npm run build-spec && node spec-client.mjs",
-    "test": "echo \"Error: no test specified\" && exit 1"
+    "test": "spago -x spec-test.dhall test"
```
Comment (hrajchert) [inline package.json:10]: This allow the test to be called from `npm t`, but we still need to add some CI configuration and call this.

File: spec-test/test/Control/JsonStream.purs [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+      stream <- liftEffect $ streamFromString ""
+      let
+        test = do
+          ev <- await
+          ev `shouldEqual` OnFinish
+
+      runProcess (test `pullFrom` jsonStreamProducer stream)
+
+    it "should call onJson on a valid json" do
+      stream <- liftEffect $ streamFromString "```{}```"
```
Comment (hrajchert) [inline spec-test/test/Control/JsonStream.purs:76]: 1 - This simple case was a bug before, when I did manual testing I always pressed an enter to end the line, but when the driver pushed the commands, some time the enter was in a different chunk and the event never fired

File: spec-test/test/Control/JsonStream.purs [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+        test = do
+          ev1 <- await
+          ev1 `shouldEqual` OnError InvalidJson
+          ev2 <- await
+          ev2 `shouldEqual` OnFinish
+
+      runProcess (test `pullFrom` jsonStreamProducer stream)
+
+    it "should recover after error" do
+      stream <- liftEffect $ streamFromString "```abc``` ```9```"
```
Comment (hrajchert) [inline spec-test/test/Control/JsonStream.purs:130]: 2 - The recover logic wasn't working as expected before

File: spec-test/src/Control/JsonStream.js [hunk 6]
```diff
@@ -15,44 +15,46 @@ export function _createJsonStream({
...
         }
 
         // If we are in Json mode, we need to see if we are ending the sequence
         // or if we are accumulating chars into the jsonBuffer (with possible resize)
         if (inJson) {
           if (
-            i + endSeparator.length < chunk.length &&
-            chunk.toString().substring(i, i + endSeparator.length) ==
-              endSeparator
+            i + endSeparator.length <= chunk.length &&
```
Comment (hrajchert) [inline spec-test/src/Control/JsonStream.js:54]: Changing `<` for `<=` solved bug number 1

File: spec-test/src/Control/JsonStream.js [hunk 7]
```diff
@@ -63,8 +65,8 @@ export function _createJsonStream({
               onError(invalidJson)();
             } else {
               onJson(json)();
-              i += endSeparator.length;
             }
+            i += endSeparator.length;
```
Comment (hrajchert) [inline spec-test/src/Control/JsonStream.js:69]: Moving the increment outside the if solved bug number 2

**Review by palas — APPROVED** [review]
Looks good. Nice tests. Maybe the splitting in chunks would be a good candidate for property based testing

**Pull request merged**